R/paste_linter.R

```diff
@@ -111,6 +123,76 @@ paste_linter <- function(allow_empty_sep = FALSE, allow_to_string = FALSE) {
     /parent::expr
   "
 
+  slash_str <- sprintf("STR_CONST[%s]", xp_text_in_table(c("'/'", '"/"')))
+  str_not_start_with_slash <-
+    "STR_CONST[not(substring(text(), 2, 1) = '/')]"
+  str_not_end_with_slash <-
+    "STR_CONST[not(substring(text(), string-length(text()) - 1, 1) = '/')]"
+  non_str <- "SYMBOL or expr"
+
+  # Type I: paste(..., sep = "/")
+  paste_file_path_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL[text() = 'paste']
+    /parent::expr
+    /parent::expr[
+      SYMBOL_SUB[text() = 'sep']/following-sibling::expr[1][{slash_str}]
+      and not(SYMBOL_SUB[text() = 'collapse'])
+    ]
+  ")
+
+  # Type II: paste0(x, "/", y, "/", z)
+  #   more specifically, paste0 need not be replaced by file.path when
+  #   (1) there are consecutive SYMBOL|expr (like paste0(x, y))
+  #   (2) a STR_CONST precedes a SYMBOL|expr, but doesn't end with /
+  #   (3) a STR_CONST follows a SYMBOL|expr, but doesn't start with /
+  #   (4) there are consecutive STR_CONSTs but neither
+  #       the first ends with / nor the second starts with /
+  #   (5) there is only one argument
+  #   (6) there is any NUM_CONST
+  #   (7) collapse is used
+  #   (8) / is the first character in an initial STR_CONST or the
+  #       last character in a terminal STR_CONST, where file.path
+  #       replacement would somewhat awkwardly look like
+  #       file.path("", ...) and/or file.path(..., "")
+  #   (9) a STR_CONST beginning or ending with >1 slashes (like "//x")
+  # use count() to exclude paste0(x)
+  # don't use starts-with() -- (1) there's no ends-with() in XPath 1.0 and
+  #   (2) we need to cater for both ' and " quotations which ^ complexity
+  # NB: XPath's substring() signature is (text, initial_index, n_characters)
+  # an expression matching any of the not(...) conditions is _not_ a path
+  paste0_file_path_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL[text() = 'paste0']
+    /parent::expr
+    /parent::expr[
+      count(expr) > 2
+      and not(
+        expr/NUM_CONST
+        or SYMBOL_SUB[text() = 'collapse']
+        or expr[2][{slash_str}]
+        or expr[last()][{slash_str}]
+        or expr/STR_CONST[
+          substring(text(), 2, 2) = '//'
```

Comment:
How robust is this with regards to raw string literals?

Reply:
Not at all! I'll file a follow-up?

Reply:
It's probably fine, will just cause false negatives in the case they are used, right?

Reply:
Right... though I think we've usually tried to be consistent about running logic on the evaluated string itself (`get_r_string()`), rather than assuming patterns on the `STR_CONST`

Reply:
#2075